src/compiler/parser.ts

```diff
@@ -556,98 +558,111 @@ module ts {
         function setDecoratorContext(val: boolean) {
             setContextFlag(val, ParserContextFlags.Decorator);
         }
-
-        function doOutsideOfContext<T>(flags: ParserContextFlags, func: () => T): T {
-            let currentContextFlags = contextFlags & flags;
-            if (currentContextFlags) {
-                setContextFlag(false, currentContextFlags);
+        
+        function setAwaitContext(val: boolean) {
+            setContextFlag(val, ParserContextFlags.Await);
+        }
+        
+        function setAsyncParameterContext(val: boolean) {
+            setContextFlag(val, ParserContextFlags.AsyncParameter);
+        }
+        
+        function doOutsideOfContext<T>(context: ParserContextFlags, func: () => T): T {
+            let setContextFlags = context & contextFlags;
+            if (setContextFlags) {
+                setContextFlag(false, setContextFlags);
                 let result = func();
-                setContextFlag(true, currentContextFlags);
+                setContextFlag(true, setContextFlags);

```

Comment:
This is intentional. I'll add documentation. `setContextFlags` here refers to the  intersection of the current context flags and the flags that we want to be temporarily cleared. As such, we clear only that subset of flags, and set them again after we have executed `func`.